Rework controllers, templates and tests

alexander-kh opened the pull request:
Changes;

- Update [sessions_controller.rb](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-d5241d488259f32ecbe2f636133e5dda)

- Add download_unclaimed_codes action in [import.rb (renamed to offeror_actions.rb)](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-6cdca9aa432d2a6fa5a70306577e7baa) and corresponding method in [offerors.rb](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-3493d3d410b6bb89f5ab228c8f955fc5)

- Rework templates common for both offerors ([home](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-4d44a4d4aeb8f73d977ff7d29b8e24d2), [view_codes](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-73b81a7bf673ca03c57eed6255a82a85)) and [profile](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-00c11f955761cf70a01418a4773dc5f3) template for the vendor
Screenshots:
    - [vendor home](http://res.cloudinary.com/dbpb81578/image/upload/v1489514013/Redeemify/Vendor_home.png)
    - [vendor home (with uploaded codes)](http://res.cloudinary.com/dbpb81578/image/upload/v1489514247/Redeemify/Vendor_home_1.png)
    - [update vendor profile](http://res.cloudinary.com/dbpb81578/image/upload/v1489514327/Redeemify/Update_vendor_profile.png)
    - [view vendor codes](http://res.cloudinary.com/dbpb81578/image/upload/v1489514406/Redeemify/View_vendor_codes.png)

- Improve test coverage for the controllers up to 100%
    - [vendors_controller_spec.rb](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-e7993e27c4aed880e65083eaf57aa093)
    - [vendor_download_codes.feature](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-3dab981e852470829e51bcd65a30536c)
    - [vendor_view_codes.feature](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-ed5dd3af5d0f169ea89892d1d51fcc03)
    - [vendor_switch_to_user.feature](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-d879ae934ca7e35d1ec1da8eb5c36be0)

- Add flow diagrams to illustrate app's behavior from four perspectives:
    - [user](http://res.cloudinary.com/dbpb81578/image/upload/v1490531686/Redeemify/User.png)
    - [admin](http://res.cloudinary.com/dbpb81578/image/upload/v1490531698/Redeemify/Admin.png)
    - [provider](http://res.cloudinary.com/dbpb81578/image/upload/v1490531704/Redeemify/Provider.png)
    - [vendor](http://res.cloudinary.com/dbpb81578/image/upload/v1490531708/Redeemify/Vendor.png)

Uzzsen commented on app/models/vendor_code.rb:
```diff
@@ -7,11 +7,8 @@ class VendorCode < ActiveRecord::Base
-    vendor.update_attribute(:unclaimCodes, vendor.unclaimCodes - 1)
+  def associate_with(user)
+    self.update(user_id: user.id, user_name: user.name, email: user.email)
   end  
```
@alexander-kh:
Hi Alex,
any particular reason to remove the adjustments of the usedCodes and  unclaimCodes attributes when a vendor code gets associated with a user ? Also, the related respec example, i.e. 'updates vendor usedCodes and unclaimCodes attributes', testing such updates in vendor_code_spec.rb, for some reason, has been removed. 
Unless the attribute updates have been refactored to a separate function (yet unknown to me), this would seem like a bug or there is a change in initial requirements.

alexander-kh replied:
Hi, Oleg
It was done intentionally.
The only code we can assign to the user is redeemify code.
To see the same offers from vendors of customers page we associate vendor codes with the user until they will be redeemed.
To illustrate on an example:
1. log in as vendor;
2. upload codes;
3. log in as a user;
4. enter valid redeemify code;
5. switch to the vendor account.
The number of vendor codes has been reduced by one, whereas it should remain unchanged.

Uzzsen replied:
Alex,
I don't see why should the number of used and unclaimed vendor codes remain unchanged for the first user login even though that user switched to the vendor account. 
Contrary, consider the case of 50 uploaded vendor codes and 60 redeemify code loggers (users). Some 10 of them will be alerted with 'Some offers are not available at this time, please come back later' which would seem right. However, the vendor home page will show 0 unclaimed codes and 50 unused codes on the bar chart thus misleading the vendor of no codes actually used an no need for additional uploads.

alexander-kh replied:
Oleg, it is a sensible note.
From my point of view, it's not correct to change the number of codes until the vendor code is redeemed.
I think we can extract this question to the separate discussion.

Uzzsen commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,135 +1,112 @@
+  def find_or_create_user
+    auth = request.env["omniauth.auth"]
+    User.find_by_provider_and_email(auth["provider"], auth["uid"]) ||
+    User.create_with_omniauth(auth)
```
This would seem like a typo:
the user is searched by_provider_and_email with auth["uid"]) for the email argument. 
The prior version featured
User.find_by_provider_and_**uid**(auth["provider"], auth["uid"]) .

alexander-kh replied:
Yes, thanks for attentiveness!

Uzzsen commented on app/models/concerns/offeror.rb:
```diff
@@ -30,20 +30,26 @@ def import(file, comment)
   
+  def download_codes(offeror_codes)
+    @unclaimed_codes = offeror_codes.where(user_id: nil)
+    @num = self.unclaimCodes
```
@alexander-kh :
For better semantics of the method name, `download_unclaimed_codes` instead of `download_codes` would  convey the intended functionality.

alexander-kh replied:
Yes, perhaps, as will be the function of downloading used codes. I kept the shorter name associating it with unclaimed codes by default and to not to make it excessively long, but as you've paid attention to this, let's change to more evident.

tansaku commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,135 +1,112 @@
-
-  def logout_user
+  
+  def set_vendor_code(current_user)
```
this methods a bit long - we should get codeclimate on here :-)

alexander-kh replied:
I agree with both statements.

tansaku commented:
definitely some improvements here - my biggest worry is that there are alot of different changes in the same PR, which makes it more challenging to review and sign off on ...

alexander-kh commented:
@tansaku,
Yes, absolutely agree.
The major changes are in templates. These are basic versions. I've made screenshots to review changes visually and not to concentrate on code.
The second goal was to improve test coverage for the controllers.
You may check results [here](https://codeclimate.com/github/alexander-kh/redeemify/coverage).
Remains changes in [sessions_controller.rb](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-d5241d488259f32ecbe2f636133e5dda)

Uzzsen commented on app/views/application/view_codes.html.erb:
```diff
@@ -0,0 +1,45 @@
+    </tr>
+  </thead>
+  <tbody>
+    <% @offeror_codes.each do |code| %>
```
With no Redemify or vendor codes persisting in the db and related checks missing in the controller action:
```
def view_codes
    @offeror_codes = offeror_codes.all
end
```
the nil set @offeror_codes will lead to a bug,

Uzzsen commented on app/controllers/concerns/import.rb:
```diff
@@ -36,16 +36,25 @@ def import
+  end
+  
+  def view_codes
+    @offeror_codes = offeror_codes.all
```
Though the View Codes button (invoking this action) is not rendered when neither Redemify nor vendor codes persist in the db, the nil object bug can be reproduced by submitting the GET 'vendors/view_codes' or GET 'providers/view_codes' request via the browser address bar and the codes missing in the persistence tier.
The codes availability check and the redirect_to statement should likely be appropriate.

alexander-kh replied:
Agree. These kinds of errors should be handled as well.

alexander-kh replied:
Oleg, I've added before_action filters as a general solution to prevent invoking unpermitted actions by constructing requests.
In the case you've mentioned, during provider or vendor session, this request will show a page without codes (offeror_codes will return an empty collection). In other cases will occur redirection to the main page.

alexander-kh replied:
@tansaku,
[set_vendor_code](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-d5241d488259f32ecbe2f636133e5ddaR98) method reworked.

alexander-kh commented:
@tansaku,
[set_vendor_code](https://github.com/strawberrycanyon/redeemify/pull/77/files#diff-d5241d488259f32ecbe2f636133e5ddaR98) method reworked.

Uzzsen commented on app/views/application/view_codes.html.erb:
```diff
@@ -0,0 +1,45 @@
+      <tr class="<%= cycle('odd_row', 'even_row') -%>">
+          <td class="width_25"><%= code.code %></td>
+          <% if current_vendor %>
+            <td class="width_10"><%= current_vendor.cashValue %></td>
```
It would probably be nicer to display the vendor-related data (cash value, expiration, description, instruction, help link) only once for the `current_vendor` and not replicate it for each `code.code` of the single vendor only.

alexander-kh replied:
This page apparently will be redesigned entirely. The goal was to create simple versions that can be easily grasped and improved.
Of course, they can be better elaborated, and they should, but let's leave it for the future iterations if you don't mind. Anyway, thank you! The proposal is good.

Uzzsen replied:
Sure, it can worked on within a separate PR.

Uzzsen commented on app/controllers/sessions_controller.rb:
```diff
@@ -1,135 +1,116 @@
+      @codes[vendor.name] ||= vendor_code ? vendor_code.code : 'Not available'
+      vendor.attributes.slice('cashValue', 'expiration', 'description',
+        'instruction', 'website', 'helpLink').each { 
+          |key, value| instance_variable_get("@#{key}")[vendor.name] ||= value }
```
@alexander-kh: 👍

alexander-kh committed:
Rework controllers, templates and tests

- Update sessions_controller.rb

- Add download_unclaimed_codes action in offeror_actions.rb

- Improve test coverage for the controllers up to 100%

- Rework templates common for both offerors (home, view_codes) and profile template for the vendor

alexander-kh committed:
Extract method for updating codes statistics

mattlindsey reviewed (approved):
I did a full QA test on the branch, and it looks good for merging to me.
